### ISSUE TITLE
Performance Issues with Postgress

### ISSUE THREAD

**Issue by vadim-kupovykh** [hangfire-postgres/Hangfire.PostgreSql#348]
Hi guys,

Here is query we see in Postgress logs
``` sql
UPDATE
  "hangfire"."jobqueue"
SET
  "fetchedat" = NOW()
WHERE
  "id" = (
  SELECT
    "id"
  FROM
    "hangfire"."jobqueue"
  WHERE
    "queue" = ANY ($1)
    AND ("fetchedat" IS NULL
      OR "fetchedat" < NOW() + INTERVAL $2)
  ORDER BY
    "fetchedat"
  NULLS FIRST
    ,
    "queue",
    "jobid" FOR
  UPDATE
    SKIP LOCKED
  LIMIT
    $3 ) RETURNING "id" AS "Id",
  "jobid" AS "JobId",
  "queue" AS "Queue",
  "fetchedat" AS "FetchedAt"
```

We have issues with performance on the Postgress side, sometimes execution up to 8 seconds of that query. We don't have so many jobs in the queue, up to 20K. Any ideas about what it can be?

**Comment by dmitry-vychikov** [comment]
Hi, 

Please share additional info:
- hangfire and hangfire postgres versions
- Hangfire configuration code
- How many workers and queues do you have
- How many hangfire processes do you run
- Dotnet version
- Is it deployed on cloud? 
- what's CPU and mem usage on the postgres server?

**Comment by vadim-kupovykh** [comment]
Hi, Sure

* Hangfire.Core - 1.8.6
* Hangfire.PostgreSql -  1.20.4.0

```
services.AddHangfire(globalConfiguration =>
            {
                globalConfiguration.UsePostgreSqlStorage(Configuration.GetConnectionString(HangfireExtensions.HangfireDatabase));
                globalConfiguration.UseSerializerSettings(HangfireExtensions.HangfireJsonSerializerSettings);
                globalConfiguration.UseTagsWithPostgreSql();
            });

services.AddHangfireServer(options =>
            {
                options.WorkerCount = agentSettings.GetValue("WorkerCount", 15);
                options.Queues = targetQueues;
            });
```

We have in GCP a few servers. Each server could have specific list of queues. Independent database for Hangfire as service in GCP. Database CPU - 100%, memory - around 50%.

![image](https://github.com/hangfire-postgres/Hangfire.PostgreSql/assets/87189559/eedbb56d-1922-4ff3-be58-05ff5d5bafba)

We can add additional instances fast. However, we have that issue when we have even 1 server. 

Execution plan of that query
![image](https://github.com/hangfire-postgres/Hangfire.PostgreSql/assets/87189559/449935c6-a439-41a6-8d96-a8ead5b20faa)

13 secs latency for index scan.

**Comment by vadim-kupovykh** [comment]
It seems like the database is feeling better when the queue is smaller. Latency better and better with decreasing queue. 
![image](https://github.com/hangfire-postgres/Hangfire.PostgreSql/assets/87189559/258c218e-6424-4f2e-9683-71cd969a50f2)

Any suggestions? :)

**Comment by dmitry-vychikov** [comment]
Try to reduce polling interval for workers. There must be a settings for this inside of in `AddHangfireServer` or `UsePostgresSqlStorage` calls.
If you have lots of workers that constantly poll the db, but never get any jobs (e.g because their own queue is empty), this can create unnecessary stress.

**Comment by vadim-kupovykh** [comment]
That is not the desired behavior. We have small jobs and all servers could process it. So, technically increasing the pooling interval (15 seconds now) will mean that workers are just doing nothing sometimes. Or am I wrong?

**Comment by azygis** [comment]
Generated a million rows in the table. [Here are the results](https://explain.dalibo.com/plan/6a7ge91abc3849ba). Even with a milion rows, query runs in sub-1s.

**Comment by vadim-kupovykh** [comment]
> Generated a million rows in the table. [Here are the results](https://explain.dalibo.com/plan/6a7ge91abc3849ba). Even with a milion rows, query runs in sub-1s.

That is my expectation. :) Will try to share the JSON plan later. I cannot it extract from GCP so simply. UseTagsWithPostgreSql enables Tags functionality by creating an independent table. It doesn't affect the initial table and selecting from jobqueue. OK, we will try to increase pooling interval.

**Comment by azygis** [comment]
I'd suggest trying manual vacuum on the table, if GCP allows that in some way. We've had some issues with autovacuum not doing the job so well in our business apps.

**Comment by vadim-kupovykh** [comment]
> I'd suggest trying manual vacuum on the table, if GCP allows that in some way. We've had some issues with autovacuum not doing the job so well in our business apps.

Tried, didn't help :(

**Comment by vadim-kupovykh** [comment]
I've found the reason. The reason that we have a lot of queue assigned to server:

```
SELECT "id"
                FROM "hangfire"."jobqueue"
                WHERE
                    "queue" = ANY (ARRAY['queue1', .... , 'queueN', 'default'])
                  AND ("fetchedat" IS NULL
                    OR "fetchedat" < NOW() + INTERVAL '1 YEAR')
                ORDER BY
                    "fetchedat" NULLS FIRST,
                    "queue",
                    "jobid"
                    FOR UPDATE SKIP LOCKED
                LIMIT 1
```

We'd like to associate specific queues with server. I see that the index on queue and fetchdat fields are created. Any ideas on how to optimize it?

**Comment by vadim-kupovykh** [comment]
So, the reason is 100% related to amount of records in jobqueu. That subquery is the issue

```
SELECT "id"
                FROM "hangfire"."jobqueue"
                WHERE
                    "queue" = ANY (ARRAY['kafka-consumption'])
                  AND ("fetchedat" IS NULL
                    OR "fetchedat" < NOW() + INTERVAL '1 YEAR')
                ORDER BY
                    "fetchedat" NULLS FIRST,
                    "queue",
                    "jobid"
                for update skip locked
                LIMIT 1;
```

Here is the plan for the query https://explain.dalibo.com/plan/ga35182eddag2f6h# 

As I understand we have problem because of ```for update skip locked```. Without ```for update skip locked``` plan: https://explain.dalibo.com/plan/76gad6c1eec5bd4e But honestly, I don't understand why the plan is so different.

**Comment by dmitry-vychikov** [comment]
https://postgrespro.com/list/thread-id/2505440 Here a similar problem is discussed.

I quickly scanned through, and one recommendation is to introduce jitter lag to polling threads to decrease contention.

Maybe there is something better, need to read until the end.

**Comment by vadim-kupovykh** [comment]
> https://postgrespro.com/list/thread-id/2505440 Here a similar problem is discussed.
> 
> I quickly scanned through, and one recommendation is to introduce jitter lag to polling threads to decrease contention.
> 
> Maybe there is something better, need to read until the end.

Checked that thread. Interesting, but I don’t understand what can we do. Will Increasing pooling interval help us? Honestly I don’t understand how to use Postgres in production now. We are doing everything by best practices, however additional workers fully destroying performance of job processing. Why nobody else has that issue. Or nobody is using Postgres in production?

**Comment by azygis** [comment]
Plenty of production applications use both PostgeSQL _and_ this package. Each environment has different capabilities, bandwidth, hardware and so on, so each environment can behave differently.

I personally have no experience with GCP so I can't really say anything about it.

**Comment by vadim-kupovykh** [comment]
> Plenty of production applications use both PostgeSQL _and_ this package. Each environment has different capabilities, bandwidth, hardware and so on, so each environment can behave differently.
> 
> I personally have no experience with GCP so I can't really say anything about it.

Yeah, just topic above is not related to hangfire and the current package, but Postgres. That is what I don’t understand why we have that issue but nobody else. What are we doing differently?:) I do see that issue related to processing “for update skip locked” in many threads. Any ideas what can we do with that?

**Comment by azygis** [comment]
Oh, yeah, it's not about Hangfire.PostgreSql storage provider, _but_ it is about PostgreSQL itself, which the provider _does_ use.

I have to admit I've not dug into the performance/indexing/etc mechanisms in PG as I had no need for it. If GCP has a limited set of operations that you can perform on the database directly (again, no idea how GCP works), my suggestion is to perhaps try [Redis provider](https://github.com/marcoCasamento/Hangfire.Redis.StackExchange). At least according to [Hangfire docs](https://docs.hangfire.io/en/latest/configuration/using-redis.html) it has a huge throughput increase compared to SQL (makes sense, since Redis is in-memory cache).

**Comment by vadim-kupovykh** [comment]
> Oh, yeah, it's not about Hangfire.PostgreSql storage provider, _but_ it is about PostgreSQL itself, which the provider _does_ use.
> 
> I have to admit I've not dug into the performance/indexing/etc mechanisms in PG as I had no need for it. If GCP has a limited set of operations that you can perform on the database directly (again, no idea how GCP works), my suggestion is to perhaps try [Redis provider](https://github.com/marcoCasamento/Hangfire.Redis.StackExchange). At least according to [Hangfire docs](https://docs.hangfire.io/en/latest/configuration/using-redis.html) it has a huge throughput increase compared to SQL (makes sense, since Redis is in-memory cache).

I believe you don’t understand the severity of that issue. Topic above is not related to GCP but yes, to something specific inside of Postgres. Until we understand what is that and how to mitigate it everyone who is using Postgres with current package under risk of performance degradation. It’s time bomb. It could happen once you try to add additional workers, or maybe not. Until we know what is that - production risks are too high for everyone.

**Comment by azygis** [comment]
I agree; to an extent.

That could happen with or without this package, as the linked mailing list shows. Without a reproduction there's virtually no way to see what can or can not happen.

I understand how you feel, being the only one which reported the issue in this repo (I'm not saying you're the only one that's impacted; just about reporting). Unfortunately, you're the one that has a reproduction and means to see whether any changes to the queries are valid or not. Since it's an OSS package, you can submit a pull request if there's a possible fix.

**Comment by dmitry-vychikov** [comment]
@vadim-kupovykh 
Can you provide a minimal runnable sample app that reproduces your issue say within a docker environment?

I may have some free time this weekend to try come up with something.

I have only 2 ideas so far:
- Introduce jitter delay within Hangfire worker, as recommended by the article above
- Try to change the the DB query in a way that sacrifices job ordering in exchange for performance gains. And expose it as a configuration option.

**Comment by azygis** [comment]
From what I understand, it's really about many workers trying to handle the same table, as in essence `SKIP LOCKED` should simply skip the locked rows (read - locked by other workers), but feels like postgresql can choke a little with _many_ rows and _many_ concurrent updates.

I'm not convinced there's a good way to just drop the ordering, even if configurable. All jobs must be picked up in order they were created.

But this gave me an idea - maybe we lack the ordered index for this particular query?

**Comment by dmitry-vychikov** [comment]
> But this gave me an idea - maybe we lack the ordered index for this particular query?

Probably not. At least, I see these 3 indexes that I think should work:
```sql
create index ix_hangfire_jobqueue_jobidandqueue
    on jobqueue (jobid, queue);

create index ix_hangfire_jobqueue_queueandfetchedat
    on jobqueue (queue, fetchedat);

create index jobqueue_queue_fetchat_jobid
    on jobqueue (queue, fetchedat, jobid);
```


# Requirements
>  All jobs must be picked up in order they were created

Can you point me where it is documented that Hangfire-compatible storage _must_ ensure ordered job pickup?

I just checked MSSQL server implementation, and they do NOT order the jobs on deque (see https://github.com/HangfireIO/Hangfire/blob/main/src/Hangfire.SqlServer/SqlServerJobQueue.cs#L191-L201)

Here's their SQL query:

```sql
set nocount on;set xact_abort on;set tran isolation level read committed;

update top (1) JQ
set FetchedAt = GETUTCDATE()
output INSERTED.Id, INSERTED.JobId, INSERTED.Queue, INSERTED.FetchedAt
from [{_storage.SchemaName}].JobQueue JQ with (forceseek, readpast, updlock, rowlock)
where Queue in @queues and
(FetchedAt is null or FetchedAt < DATEADD(second, @timeoutSs, GETUTCDATE()));
```

Ordering question has been discussed here https://github.com/HangfireIO/Hangfire/issues/398 , and it seems that the correct way to maintain job dependencies is using job continuation.

# Usage scenarios
I think that it really depends on the use case of each consumer whether ordering is important or not.
Some users (just as myself) suffered from enforced ordering pattern. We had 4 queues, and wanted them to be processed almost evenly, but it was not possible because workers would pickup jobs using order by queue names.

Introducing an option to avoid sorting can be considered a feature that gives users even more control over their processing pipeline.

# Performance
Ensuring ordering is really putting much stress on performance.
I did a very stupid investigation, but results can be seen easily.
I generated 4 queues with 30_000 jobs each, and ran hangfire server two times:
- w/ ordering
    -  consumed 1600% of my CPU (all cores full load)
    - only 20,000 jobs were processed within approx. 3 minutes
- w/o ordering
    - was consuming about 600% (almost 3 times less)
    - All 120,000 jobs processed within 3 minutes

See the container CPU usage chart from docker.
![image](https://github.com/hangfire-postgres/Hangfire.PostgreSql/assets/31896999/4100bb69-e5b2-4834-9c84-01c36ecd1f5e)



## My setup
- Windows 11, x64
- I7-10700k (16 cores)
- 48GB RAM
- latest postgres, running in docker
- 1 hangfire server process, 60 workers
- Empty jobs that do nothing:
    - ```c#
         public async Task Delayed()
         {}
      ```


# Suggested next steps
@azygis I stand firm on my suggestion with a config flag that removes ordering in exchange for more performance.
If you approve, I can try to submit a PR sometime next week.

@vadim-kupovykh do you think this tradeoff would be applicable for your case, at least?


Though, it may require additional tweaks on Dashboard and other parts of the library. On the dashboard, I saw strange artifacts when page with queue jobs would show "No jobs in queue" occasionally during active processing.

# Useful reading
I also read this article. It is basically saying that `FOR UPDATE SKIP LOCKED` is the best way to do queues in postgres, but performance will suffer with multiple workers.

https://www.2ndquadrant.com/en/blog/what-is-select-skip-locked-for-in-postgresql-9-5/

# Source code of testing app

```c#
using Hangfire;
using Hangfire.PostgreSql;

string[] queues = { "alpha", "beta", "gamma", "test" };

bool generateJobs = true; //Toggle to switch between creating new jobs and processing them

var builder = WebApplication.CreateBuilder(args);

builder.Services.AddHangfire(configuration => {
  configuration.UsePostgreSqlStorage(options => {
    options.UseNpgsqlConnection("User ID=postgres;Password=mysecretpassword;Host=localhost;Port=5432;Database=hangfire-postgres");
  });
});


if (!generateJobs)
{
  builder.Services.AddHangfireServer(options => {
    options.Queues = queues;
    options.WorkerCount = 60;
  });
}

var app = builder.Build();

if (generateJobs)
{
  var client = app.Services.GetRequiredService<IBackgroundJobClientV2>();

  foreach (string queue in queues)
  {
    Task.Run(() => {
      for (int i = 0; i < 30_000; i++)
      {
        client.Enqueue<Jobs>(queue, jobs => jobs.Delayed());
      }
    });
  }
}

app.UseHangfireDashboard("");
app.Run();


public class Jobs
{
  public async Task Delayed()
  {
  }
}
```
## Toggling job ordering
I compiled Hangfire.Postgres locally, and basically removed `ORDER BY ...` statement from here: https://github.com/hangfire-postgres/Hangfire.PostgreSql/blob/ffb935257cd006ec9e01df536ef98005904faec0/src/Hangfire.PostgreSql/PostgreSqlJobQueue.cs#L143

**Comment by azygis** [comment]
What if you specify [index order](https://www.postgresql.org/docs/current/indexes-ordering.html)? Currently it's a default, which is ASC NULLS LAST. Query does a complete 180 with NULLS FIRST.

```
create index jobqueue_queue_fetchat_jobid
    on jobqueue (fetchedat NULLS FIRST, queue, jobid);
```
Something like that?

> Can you point me where it is documented that Hangfire-compatible storage must ensure ordered job pickup?

I cannot. It's a mistake done by many, and since the package has been with this ordering for quite some time, it can become a breaking change for some people. Of course, with a default of "keep ordering" it's not a breaking change though.

**Comment by dmitry-vychikov** [comment]
> What if you specify [index order](https://www.postgresql.org/docs/current/indexes-ordering.html)? Currently it's a default, which is ASC NULLS LAST. Query does a complete 180 with NULLS FIRST.

@azygis brilliant idea!

On my test data it works as fast as w/o ordering.

**Comment by vadim-kupovykh** [comment]
Guys, you are amazing.:) going to test with ordered index! Will update you later.

**Comment by azygis** [comment]
Lovely! Then we don't need any conditionals, I suppose?

Would you be able to look at other indexes? If order is needed for any of them?

**Comment by dmitry-vychikov** [comment]
> Lovely! Then we don't need any conditionals, I suppose?

Yes, we don't. I tried to test two cases with regards to ordering:
1) Reversed order by queue name when query has to pick first jobs from the bottom of the queue
2) Evenly distributed jobs based on queue name.

I didn't see any significant difference between these two. Both cases run faster with updated index. At least, we did not make it worse.

> Would you be able to look at other indexes? If order is needed for any of them?

I might take a look next weekend.

Right now I'm not even sure if other indexes for `jobqueue` are needed at all. I had a brief look at past commits. , and there were performance improvements creating new indexes in that table, but old ones were not touched.

Maybe it's time to do some clean ups.

**Comment by vadim-kupovykh** [comment]
@dmitry-vychikov I've checked on my side with just applying index order - so, we don't have that unexpected degradation anymore. The query is executing ten times faster. Please merge into master, we are ready to update ASAP on all our environments. 

We are continuing to monitor that the issue will not reproduce, but here is the latency before applying index order and after:
![image](https://github.com/hangfire-postgres/Hangfire.PostgreSql/assets/87189559/57e94b91-6a43-44f6-951a-60dd7924165e)

I hope we will not have degradation anymore.

@dmitry-vychikov, @azygis thank you guys for the fast solution!

**Comment by azygis** [comment]
Published as 1.20.8.
